browserdebuggertools/chrome/interface.py

```diff
@@ -185,9 +182,121 @@ def get_opened_javascript_dialog(self):
         """
         Gets the opened javascript dialog.
 
-        :raises DomainNotFoundError: If the Page domain isn't enabled
+        :raises DomainNotEnabledError: If the Page domain isn't enabled
         :raises JavascriptDialogNotFoundError: If there is currently no dialog open
         """
         return (
             self._socket_handler.event_handlers["JavascriptDialog"].get_opened_javascript_dialog()
         )
+
+    def get_iframe_source_content(self, xpath):
+        # type: (str) -> str
+        """
+        Returns the HTML markup for an iframe document, where the iframe node can be located in the
+        DOM with the given xpath.
+
+        :param xpath: following the spec 3.1 https://www.w3.org/TR/xpath-31/
+        :return: HTML markup
+        :raises XPathSyntaxError: The given xpath is invalid
+        :raises IFrameNotFoundError: A matching iframe document could not be found
+        :raises UnknownError: The socket handler received a message with an unknown error code
+        """
+
+        try:
+            XPath(xpath)  # Validates the xpath
+
+        except XPathSyntaxError:
+            raise InvalidXPathError("{0} is not a valid xpath".format(xpath))
+
+        return self._get_iframe_html(xpath)
+
+    def get_page_source(self):
+        # type: () -> str
+        """
+        Returns the HTML markup of the current page. Iframe tags are included but the enclosed
+        documents are not. Consider enabling the Page domain to increase performance.
+
+        :return: HTML markup
+        """
+
+        root_node_id = self._socket_handler.event_handlers["PageLoad"].get_root_node_id()
+        return self._get_outer_html(root_node_id)
+
+    def _get_iframe_html(self, xpath, _attempts=0):
+        # type: (str, int) -> str
+
+        try:
+            backend_node_id = self._get_iframe_backend_node_id(xpath)
+            return self._get_outer_html(backend_node_id)
+
+        except ResourceNotFoundError:
+            # The cached node doesn't exist any more, so we need to find a new one that matches
+            # the xpath. Backend node IDs are unique, so there is not a risk of getting the
+            # outer html of the wrong node.
+            if _attempts < 1:
+                if xpath in self._node_map:
+                    del self._node_map[xpath]
+                return self._get_iframe_html(xpath, _attempts=_attempts + 1)
```

Comment:
This could be done in a loop to avoid the stack from increasing. It shortens the stack trace and saves on memory slightly.

Reply:
Or even unroll the loop completely since we never want to try more than once. Because either the node is there or not.